src/persist-client/src/internal/metrics.rs

```diff
@@ -1654,6 +1670,185 @@ impl SinkMetrics {
+            correction_sum_capacity_decreases: self.correction_sum_capacity_decreases.clone(),
+            correction_previous_cap: 0,
+            correction_max_capacity_per_worker,
+        }
+    }
+
+    /// Updates our estimate of the aggregate peak for the correction buffer
+    /// length and capacity across workers and persist sink.
+    ///
+    /// This method assumes temporary quiescence on the correction buffer metrics,
+    /// i.e., there are no updates to these metrics taking place when the method
+    /// is called. This occurs, e.g., after a `step_or_park` call.
+    pub fn update_sink_correction_peak_metrics(&self) {
+        // Correctness argument:
+        // 1. We always update insertions/increases prior to updating deletions/decreases
+        // in each worker. So upon quiescence, the net effect of these must result in a
+        // non-negative quantity.
+        // 2. Every worker will execute the instructions below and compute the same peak.
+        // This is because during quiescence, no changes happen to the previous peak nor
+        // other metrics, and the ones we look at are all shared across workers. So at least
+        // one worker will update the peak with its new value, if there is an increase, at
+        // the quiescence granularity.
+        // Note that we may miss the overall peak if between quiescence moments there are
+        // lots of insertions followed by lots of deletions to the correction buffer. We
+        // find this trade-off more attractive than adding explicity multi-metric synchronization
+        // among workers.
+        let aggregate_correction_len = self
+            .correction_sum_insertions
+            .get()
+            .checked_sub(self.correction_sum_deletions.get());
+        let Some(aggregate_correction_len) = aggregate_correction_len else {
+            error!(
+                aggregate_correction_len,
+                "Negative aggregate length for persist sink correction"
+            );
+            return;
+        };
+        if aggregate_correction_len > self.correction_peak_len.get() {
+            self.correction_peak_len.set(aggregate_correction_len);
+        }
+        let aggregate_correction_cap = self
+            .correction_sum_capacity_increases
+            .get()
+            .checked_sub(self.correction_sum_capacity_decreases.get());
+        let Some(aggregate_correction_cap) = aggregate_correction_cap else {
+            error!(
+                aggregate_correction_cap,
+                "Negative aggregate capacity for persist sink correction"
+            );
+            return;
+        };
+        if aggregate_correction_cap > self.correction_peak_capacity.get() {
+            self.correction_peak_capacity.set(aggregate_correction_cap);
+        }
+    }
+}
+
+/// Metrics for the persist sink that are labeled per-worker.
+#[derive(Clone, Debug)]
+pub struct SinkWorkerMetrics {
```

Comment:
This struct seems odd to me. It is a subset of sink metrics, but not only per worker, but also per persist sink. This mixture isn't obvious and as such it seems to be easy to break in the future. With the approach mentioned in the comment, the non-gauge parts of the struct could go away.

Reply:
I like your suggestion with the `CorrectionBuffer` struct, so I adopted it with some small adaptations to handle `usize` differences. As part of that change, I also restricted the `SinkWorkerMetrics` struct to only per-worker metrics, and separated the reports of metrics for `SinkMetrics` from the ones for `SinkWorkerMetrics`.